File: backend/src/AgentTeam/Models/Agent.php
<?php

declare(strict_types=1);

namespace AgentTeam\Models;

/**
 * Agent Model
 *
 * Represents an AI agent with its configuration, tools, and settings.
 * Supports three types: standard, manager, and worker.
 *
 * @see docs/agentDesign.md
 */
class Agent
{
    private ?int $id = null;
    private int $userId = 0;
    private ?int $teamId = null;
    private ?string $category = null;
    private string $name = '';
    private string $description = '';

    // Agent Type & Hierarchy
    private string $agentType = 'standard';  // standard, manager, worker
    private ?int $parentAgentId = null;
    private array $canDelegateTo = [];
    private int $displayOrder = 0;  // For drag-and-drop ordering within team

    // LLM Configuration
    private string $provider = 'claude';
    private ?string $model = null;
    private string $instructions = '';

    // Tools
    private array $tools = [];

    // Access Control
    private string $visibility = 'personal';  // personal, workspace, public
    private bool $enabled = true;

    // Settings
    private array $settings = [];

    // Timestamps
    private ?string $createdAt = null;
    private ?string $updatedAt = null;

    /**
     * Create a new Agent instance
     */
    public function __construct(array $data = [])
    {
        if (!empty($data)) {
            $this->hydrate($data);
        }
    }

    /**
     * Hydrate the agent from an array (e.g., database row)
     */
    public function hydrate(array $data): self
    {
        $this->id = isset($data['id']) ? (int) $data['id'] : null;
        $this->userId = isset($data['user_id']) ? (int) $data['user_id'] : 0;
        $this->teamId = isset($data['team_id']) ? (int) $data['team_id'] : null;
        $this->category = isset($data['category']) && $data['category'] !== '' ? (string) $data['category'] : null;
        $this->name = $data['name'] ?? '';
        $this->description = $data['description'] ?? '';

        $this->agentType = $data['agent_type'] ?? 'standard';
        $this->parentAgentId = isset($data['parent_agent_id']) ? (int) $data['parent_agent_id'] : null;
        $this->canDelegateTo = $this->decodeJson($data['can_delegate_to'] ?? []);
        $this->displayOrder = isset($data['display_order']) ? (int) $data['display_order'] : 0;

        $this->provider = $data['provider'] ?? 'claude';
        $this->model = $data['model'] ?? null;
        $this->instructions = $data['instructions'] ?? '';

        $this->tools = $this->decodeJson($data['tools'] ?? []);

        $this->visibility = $data['visibility'] ?? 'personal';
        $this->enabled = (bool) ($data['enabled'] ?? true);

        $this->settings = $this->decodeJson($data['settings'] ?? []);

        $this->createdAt = $data['created_at'] ?? null;
        $this->updatedAt = $data['updated_at'] ?? null;

        return $this;
    }

    /**
     * Decode JSON string or return array as-is
     */
    private function decodeJson($value): array
    {
        if (is_string($value) && !empty($value)) {
            $decoded = json_decode($value, true);
            return is_array($decoded) ? $decoded : [];
        }
        return is_array($value) ? $value : [];
    }

    /**
     * Convert the agent to an array
     */
    public function toArray(): array
    {
        return [
            'id' => $this->id,
            'user_id' => $this->userId,
            'team_id' => $this->teamId,
            'category' => $this->category,
            'name' => $this->name,
            'description' => $this->description,
            'agent_type' => $this->agentType,
            'parent_agent_id' => $this->parentAgentId,
            'can_delegate_to' => $this->canDelegateTo,
            'display_order' => $this->displayOrder,
            'provider' => $this->provider,
            'model' => $this->model,
            'instructions' => $this->instructions,
            'tools' => $this->tools,
            'visibility' => $this->visibility,
            'enabled' => $this->enabled,
            'settings' => $this->settings,
            'created_at' => $this->createdAt,
            'updated_at' => $this->updatedAt,
        ];
    }

    /**
     * Convert to array for JSON API response (public-safe fields)
     */
    public function toApiArray(): array
    {
        return [
            'id' => $this->id,
            'name' => $this->name,
            'description' => $this->description,
            'category' => $this->category,
            'agent_type' => $this->agentType,
            'display_order' => $this->displayOrder,
            'provider' => $this->provider,
            'model' => $this->model,
            'instructions' => $this->instructions,
            'tools' => $this->tools,
            'settings' => $this->settings,
            'visibility' => $this->visibility,
            'enabled' => $this->enabled,
            'created_at' => $this->createdAt,
        ];
    }

    /**
     * Build the system prompt for this agent
     */
    public function buildSystemPrompt(): string
    {
        $prompt = "You are {$this->name}.";

        if (!empty($this->description)) {
            $prompt .= "\n\n{$this->description}";
        }

        if (!empty($this->instructions)) {
            $prompt .= "\n\n## Instructions\n{$this->instructions}";
        }

        // Add delegation info for manager agents
        if ($this->isManager()) {
            $prompt .= "\n\n## Agent Capabilities\n";
            $prompt .= "You are a manager agent with the ability to delegate tasks to specialized worker agents.\n";
            $prompt .= "Use the `list_available_agents` tool to see your team.\n";
            $prompt .= "Use `delegate_to_agent` to assign tasks to specific agents.\n";
            $prompt .= "Use `run_agents_parallel` to run multiple agents simultaneously.";
        }

        return $prompt;
    }

    /**
     * Check if this is a manager agent
     */
    public function isManager(): bool
    {
        return $this->agentType === 'manager';
    }

    /**
     * Check if this is a worker agent
     */
    public function isWorker(): bool
    {
        return $this->agentType === 'worker';
    }

    /**
     * Check if this is a standard agent
     */
    public function isStandard(): bool
    {
        return $this->agentType === 'standard';
    }

    /**
     * Check if this manager can delegate to a specific agent
     */
    public function canDelegateToAgent(int $agentId): bool
    {
        if (!$this->isManager()) {
            return false;
        }

        // Empty array means can delegate to any agent
        if (empty($this->canDelegateTo)) {
            return true;
        }

        return in_array($agentId, $this->canDelegateTo, true);
    }

    /**
     * Check if a user can access this agent
     */
    public function isAccessibleBy(int $userId): bool
    {
        // Owner always has access
        if ($this->userId === $userId) {
            return true;
        }

        // Public agents are accessible to all
        if ($this->visibility === 'public') {
            return true;
        }

        return false;
    }

    /**
     * Get a specific setting value
     */
    public function getSetting(string $key, $default = null)
    {
        return $this->settings[$key] ?? $default;
    }

    /**
     * Set a specific setting value
     */
    public function setSetting(string $key, $value): self
    {
        $this->settings[$key] = $value;
        return $this;
    }

    // ========================================
    // Getters
    // ========================================

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getUserId(): int
    {
        return $this->userId;
    }

    public function getTeamId(): ?int
    {
        return $this->teamId;
    }

    public function getCategory(): ?string
    {
        return $this->category;
    }

    public function getName(): string
    {
        return $this->name;
    }

    public function getDescription(): string
    {
        return $this->description;
    }

    public function getAgentType(): string
    {
        return $this->agentType;
    }

    public function getParentAgentId(): ?int
    {
        return $this->parentAgentId;
    }

    public function getCanDelegateTo(): array
    {
        return $this->canDelegateTo;
    }

    public function getDisplayOrder(): int
    {
        return $this->displayOrder;
    }

    public function getProvider(): string
    {
        return $this->provider;
    }

    public function getModel(): ?string
    {
        return $this->model;
    }

    public function getInstructions(): string
    {
        return $this->instructions;
    }

    public function getTools(): array
    {
        return $this->tools;
    }

    public function getVisibility(): string
    {
        return $this->visibility;
    }

    public function isEnabled(): bool
    {
        return $this->enabled;
    }

    public function getSettings(): array
    {
        return $this->settings;
    }

    public function getCreatedAt(): ?string
    {
        return $this->createdAt;
    }

    public function getUpdatedAt(): ?string
    {
        return $this->updatedAt;
    }

    // ========================================
    // Setters (Fluent Interface)
    // ========================================

    public function setId(int $id): self
    {
        $this->id = $id;
        return $this;
    }

    public function setUserId(int $userId): self
    {
        $this->userId = $userId;
        return $this;
    }

    public function setTeamId(?int $teamId): self
    {
        $this->teamId = $teamId;
        return $this;
    }

    public function setCategory(?string $category): self
    {
        $this->category = ($category === null || $category === '') ? null : $category;
        return $this;
    }

    public function setName(string $name): self
    {
        $this->name = $name;
        return $this;
    }

    public function setDescription(string $description): self
    {
        $this->description = $description;
        return $this;
    }

    public function setAgentType(string $agentType): self
    {
        if (!in_array($agentType, ['standard', 'manager', 'worker'], true)) {
            throw new \InvalidArgumentException("Invalid agent type: {$agentType}");
        }
        $this->agentType = $agentType;
        return $this;
    }

    public function setParentAgentId(?int $parentAgentId): self
    {
        $this->parentAgentId = $parentAgentId;
        return $this;
    }

    public function setCanDelegateTo(array $agentIds): self
    {
        $this->canDelegateTo = array_map('intval', $agentIds);
        return $this;
    }

    public function setProvider(string $provider): self
    {
        $this->provider = $provider;
        return $this;
    }

    public function setModel(?string $model): self
    {
        $this->model = $model;
        return $this;
    }

    public function setInstructions(string $instructions): self
    {
        $this->instructions = $instructions;
        return $this;
    }

    public function setTools(array $tools): self
    {
        $this->tools = $tools;
        return $this;
    }

    public function addTool(string $tool): self
    {
        if (!in_array($tool, $this->tools, true)) {
            $this->tools[] = $tool;
        }
        return $this;
    }

    public function removeTool(string $tool): self
    {
        $this->tools = array_values(array_filter(
            $this->tools,
            fn($t) => $t !== $tool
        ));
        return $this;
    }

    public function setDisplayOrder(int $order): self
    {
        $this->displayOrder = $order;
        return $this;
    }

    public function setVisibility(string $visibility): self
    {
        if (!in_array($visibility, ['personal', 'workspace', 'public'], true)) {
            throw new \InvalidArgumentException("Invalid visibility: {$visibility}");
        }
        $this->visibility = $visibility;
        return $this;
    }

    public function setEnabled(bool $enabled): self
    {
        $this->enabled = $enabled;
        return $this;
    }

    public function setSettings(array $settings): self
    {
        $this->settings = $settings;
        return $this;
    }
}

File: backend_python/app/agent_team/models/agent.py
"""Port of backend/src/AgentTeam/Models/Agent.php.

Represents an AI agent with its configuration, tools, and settings.
Supports three types: standard, manager, and worker.
"""
from __future__ import annotations

import json

from app.support.phpcompat import php_array, php_bool, php_empty, php_intval, php_strval


def _decode_json(value):
    """Mirror Agent::decodeJson (PHP 96-103): a non-empty string is JSON-decoded
    (invalid JSON / non-array result -> []); anything else passes through if it
    is already an array (dict/list), else []."""
    if isinstance(value, str) and not php_empty(value):
        try:
            decoded = json.loads(value)
        except ValueError:
            decoded = None
        return decoded if isinstance(decoded, (list, dict)) else []
    return value if isinstance(value, (list, dict)) else []


AGENT_TYPES = ('standard', 'manager', 'worker', 'dispatcher', 'playbook')
VISIBILITIES = ('personal', 'workspace', 'public')


class Agent:
    def __init__(self, data: dict | None = None):
        """PHP: `if (!empty($data)) { $this->hydrate($data); }` — an empty/None
        dict leaves every field at its declared default (below), never calling
        hydrate at all."""
        self.id: int | None = None
        self.userId: int = 0
        self.teamId: int | None = None
        self.category: str | None = None
        self.name: str = ''
        self.description: str = ''

        self.agentType: str = 'standard'
        self.parentAgentId: int | None = None
        self.canDelegateTo: list = []
        self.displayOrder: int = 0

        self.provider: str = 'claude'
        self.model: str | None = None
        self.instructions: str = ''

        self.tools: list = []

        self.visibility: str = 'personal'
        self.enabled: bool = True

        self.settings: dict | list = []

        self.createdAt: str | None = None
        self.updatedAt: str | None = None

        if data:
            self.hydrate(data)

    def hydrate(self, data: dict) -> 'Agent':
        self.id = php_intval(data['id']) if data.get('id') is not None else None
        self.userId = php_intval(data['user_id']) if data.get('user_id') is not None else 0
        self.teamId = php_intval(data['team_id']) if data.get('team_id') is not None else None
        category = data.get('category')
        self.category = php_strval(category) if (category is not None and category != '') else None
        self.name = data.get('name') if data.get('name') is not None else ''
        self.description = data.get('description') if data.get('description') is not None else ''

        self.agentType = data.get('agent_type') if data.get('agent_type') is not None else 'standard'
        self.parentAgentId = php_intval(data['parent_agent_id']) if data.get('parent_agent_id') is not None else None
        can_delegate_to = data.get('can_delegate_to')
        self.canDelegateTo = _decode_json(can_delegate_to if can_delegate_to is not None else [])
        self.displayOrder = php_intval(data['display_order']) if data.get('display_order') is not None else 0

        self.provider = data.get('provider') if data.get('provider') is not None else 'claude'
        self.model = data.get('model') if data.get('model') is not None else None
        self.instructions = data.get('instructions') if data.get('instructions') is not None else ''

        tools = data.get('tools')
        self.tools = _decode_json(tools if tools is not None else [])

        self.visibility = data.get('visibility') if data.get('visibility') is not None else 'personal'
        enabled = data.get('enabled')
        self.enabled = php_bool(enabled if enabled is not None else True)

        settings = data.get('settings')
        self.settings = _decode_json(settings if settings is not None else [])

        self.createdAt = data.get('created_at') if data.get('created_at') is not None else None
        self.updatedAt = data.get('updated_at') if data.get('updated_at') is not None else None

        return self

    def toArray(self) -> dict:
        return {
            'id': self.id,
            'user_id': self.userId,
            'team_id': self.teamId,
            'category': self.category,
            'name': self.name,
            'description': self.description,
            'agent_type': self.agentType,
            'parent_agent_id': self.parentAgentId,
            'can_delegate_to': php_array(self.canDelegateTo),
            'display_order': self.displayOrder,
            'provider': self.provider,
            'model': self.model,
            'instructions': self.instructions,
            'tools': php_array(self.tools),
            'visibility': self.visibility,
            'enabled': self.enabled,
            'settings': php_array(self.settings),
            'created_at': self.createdAt,
            'updated_at': self.updatedAt,
        }

    def toApiArray(self) -> dict:
        return {
            'id': self.id,
            'name': self.name,
            'description': self.description,
            'category': self.category,
            'agent_type': self.agentType,
            'display_order': self.displayOrder,
            'provider': self.provider,
            'model': self.model,
            'instructions': self.instructions,
            'tools': php_array(self.tools),
            'settings': php_array(self.settings),
            'visibility': self.visibility,
            'enabled': self.enabled,
            'created_at': self.createdAt,
        }

    def buildSystemPrompt(self) -> str:
        prompt = f"You are {self.name}."

        if not php_empty(self.description):
            prompt += f"\n\n{self.description}"

        if not php_empty(self.instructions):
            prompt += f"\n\n## Instructions\n{self.instructions}"

        if self.isManager():
            prompt += "\n\n## Agent Capabilities\n"
            prompt += "You are a manager agent with the ability to delegate tasks to specialized worker agents.\n"
            prompt += "Use the `list_available_agents` tool to see your team.\n"
            prompt += "Use `delegate_to_agent` to assign tasks to specific agents.\n"
            prompt += "Use `run_agents_parallel` to run multiple agents simultaneously."

        return prompt

    def isManager(self) -> bool:
        return self.agentType == 'manager'

    def isWorker(self) -> bool:
        return self.agentType == 'worker'

    def isStandard(self) -> bool:
        return self.agentType == 'standard'

    def canDelegateToAgent(self, agent_id: int) -> bool:
        if not self.isManager():
            return False

        # Empty array means can delegate to any agent
        if php_empty(self.canDelegateTo):
            return True

        return agent_id in self.canDelegateTo

    def isAccessibleBy(self, user_id: int) -> bool:
        # Owner always has access
        if self.userId == user_id:
            return True

        # Public agents are accessible to all
        if self.visibility == 'public':
            return True

        return False

    def getSetting(self, key: str, default=None):
        if isinstance(self.settings, dict):
            value = self.settings.get(key)
            return value if value is not None else default
        return default

    def setSetting(self, key: str, value) -> 'Agent':
        if not isinstance(self.settings, dict):
            self.settings = {}
        self.settings[key] = value
        return self

    # ========================================
    # Getters
    # ========================================

    def getId(self) -> int | None:
        return self.id

    def getUserId(self) -> int:
        return self.userId

    def getTeamId(self) -> int | None:
        return self.teamId

    def getCategory(self) -> str | None:
        return self.category

    def getName(self) -> str:
        return self.name

    def getDescription(self) -> str:
        return self.description

    def getAgentType(self) -> str:
        return self.agentType

    def getParentAgentId(self) -> int | None:
        return self.parentAgentId

    def getCanDelegateTo(self) -> list:
        return self.canDelegateTo

    def getDisplayOrder(self) -> int:
        return self.displayOrder

    def getProvider(self) -> str:
        return self.provider

    def getModel(self) -> str | None:
        return self.model

    def getInstructions(self) -> str:
        return self.instructions

    def getTools(self) -> list:
        return self.tools

    def getVisibility(self) -> str:
        return self.visibility

    def isEnabled(self) -> bool:
        return self.enabled

    def getSettings(self) -> dict:
        return self.settings

    def getCreatedAt(self) -> str | None:
        return self.createdAt

    def getUpdatedAt(self) -> str | None:
        return self.updatedAt

    # ========================================
    # Setters (Fluent Interface)
    # ========================================

    def setId(self, id: int) -> 'Agent':
        self.id = id
        return self

    def setUserId(self, user_id: int) -> 'Agent':
        self.userId = user_id
        return self

    def setTeamId(self, team_id: int | None) -> 'Agent':
        self.teamId = team_id
        return self

    def setCategory(self, category: str | None) -> 'Agent':
        self.category = None if (category is None or category == '') else category
        return self

    def setName(self, name: str) -> 'Agent':
        self.name = name
        return self

    def setDescription(self, description: str) -> 'Agent':
        self.description = description
        return self

    def setAgentType(self, agent_type: str) -> 'Agent':
        if agent_type not in AGENT_TYPES:
            raise ValueError(f'Invalid agent type: {agent_type}')
        self.agentType = agent_type
        return self

    def setParentAgentId(self, parent_agent_id: int | None) -> 'Agent':
        self.parentAgentId = parent_agent_id
        return self

    def setCanDelegateTo(self, agent_ids: list) -> 'Agent':
        self.canDelegateTo = [php_intval(a) for a in agent_ids]
        return self

    def setProvider(self, provider: str) -> 'Agent':
        self.provider = provider
        return self

    def setModel(self, model: str | None) -> 'Agent':
        self.model = model
        return self

    def setInstructions(self, instructions: str) -> 'Agent':
        self.instructions = instructions
        return self

    def setTools(self, tools: list) -> 'Agent':
        self.tools = tools
        return self

    def addTool(self, tool: str) -> 'Agent':
        if tool not in self.tools:
            self.tools.append(tool)
        return self

    def removeTool(self, tool: str) -> 'Agent':
        self.tools = [t for t in self.tools if t != tool]
        return self

    def setDisplayOrder(self, order: int) -> 'Agent':
        self.displayOrder = order
        return self

    def setVisibility(self, visibility: str) -> 'Agent':
        if visibility not in VISIBILITIES:
            raise ValueError(f'Invalid visibility: {visibility}')
        self.visibility = visibility
        return self

    def setEnabled(self, enabled: bool) -> 'Agent':
        self.enabled = enabled
        return self

    def setSettings(self, settings: dict) -> 'Agent':
        self.settings = settings
        return self
